scalacheck/src/main/scala/karat/scalacheck/Scalacheck.scala

```diff
@@ -41,4 +43,39 @@ object Scalacheck {
     )
     if (problem == null) Prop.passed else Prop(problem.getError)
   }
-}
\ No newline at end of file
+
+  def checkFormula[F[_] : Monad, Action, State, Response](actions: List[Action], initial: F[State], step: (Action, State) => F[Step[State, Response]])(
+    formula: Formula[Info[Action, State, Response]]
+  ): F[Prop] =
+    for {
+      current <- initial
+      result <- checkFormula(new ScalacheckStepResultManager[Info[Action, State, Response]](), step, actions, current, formula)
+    } yield result
+
+  def checkFormula[F[_] : Monad, Action, State, Response](
+    resultManager: ScalacheckStepResultManager[Info[Action, State, Response]],
+    step: (Action, State) => F[Step[State, Response]],
+    actions: List[Action],
+    current: State,
+    formula: Formula[Info[Action, State, Response]]
+  ): F[Prop] = Monad[F].tailRecM((actions, current, formula)) { case (actions, current, formula) =>
+    actions match {
+      case Nil =>
+        resultToProp(CheckKt.leftToProve(resultManager, formula)).asRight.pure
+      case action :: rest => step(action, current).flatMap {
+        case null =>
```

Comment:
We should find a way of managing this differently in Scala

Reply:
What do you refer to exactly?

If it's the `null`, I actually moved to nullable types in Kotlin because it's the easiest way to interoperate between JVM languages. Trying to use something more complex like `Either` was difficult in the Scala side.

Reply:
From a Scala user's perspective, do I need to return `null` when I want to stop?